python/sglang/srt/distributed/device_communicators/custom_all_reduce.py

```diff
@@ -0,0 +1,309 @@
+import ctypes
+import logging
+from contextlib import contextmanager
+from typing import List, Optional, Union
+
+import torch
+import torch.distributed as dist
+import vllm.envs as envs
```

Comment:
`envs`, `current_platform` and `cuda_device_count_stateless` should also be replaced